misk-aws/src/main/kotlin/misk/jobqueue/sqs/SqsJobConsumer.kt

```diff
@@ -63,73 +71,120 @@ internal class SqsJobConsumer @Inject internal constructor(
   }
 
   internal inner class QueueReceiver(
-    private val queueName: QueueName,
+    queueName: QueueName,
     private val handler: JobHandler
   ) {
     private val queue = queues[queueName]
 
-    fun runOnce(): Status {
-      val (receiveDuration, messages) = queue.call { client ->
-        val receiveRequest = ReceiveMessageRequest()
-          .withAttributeNames("All")
-          .withMessageAttributeNames("All")
-          .withQueueUrl(queue.url)
-          .withMaxNumberOfMessages(config.message_batch_size)
+    fun run(): Status {
+      // Receive messages in parallel. Default to 1 if feature flag is not defined.
+      val numReceivers = receiversForQueue()
+      val futures = (1..numReceivers)
+          .filter { num ->
+            val lease = leaseManager.requestLease("sqs-job-consumer-${queue.name.value}-$num")
+            !config.clustered_consumers || lease.checkHeld()
+          }
+          .map {
+            CompletableFuture.supplyAsync(Supplier {
+              receive()
+            }, receivingThreads)
+          }
 
-        timed { client.receiveMessage(receiveRequest).messages }
+      // Either all messages are consumed and processed successfully, or we signal failure.
+      // If none of the received consume any messages, return NO_WORK for backoff.
+      return CompletableFuture.allOf(*futures.toTypedArray()).thenApply {
+        futures.flatMap { it.join() }.minBy {
+          when (it) {
+            Status.FAILED -> 1
+            Status.OK -> 2
+            Status.NO_WORK -> 3
+            else -> throw IllegalStateException("Consumer returned an unexpected Status")
+          }
+        } ?: Status.NO_WORK
+      }.join()
+    }
+
+    private fun receiversForQueue(): Int {
+      return try {
+        max(featureFlags.getInt(CONSUMERS_PER_QUEUE, queue.queueName), CONSUMERS_PER_QUEUE_MIN)
+      } catch (e: Exception) {
+        log.warn { "jobqueue-consumers not defined for ${queue.queueName}" }
+        CONSUMERS_PER_QUEUE_DEFAULT
```

Comment:
why not just blow up? not a fan of using some default value

Reply:
I don't have a strong preference. I can change it back.